feat: Elevated Rate Limits core mechanism (#1)
ecita/limitd-redis: README.md

});
```

### Options available:

- `uri` (string): Redis Connection String.
- `nodes` (array): [Redis Cluster Configuration](https://github.com/luin/ioredis#cluster).
- `buckets` (object): Setup your bucket types.
- `prefix` (string): Prefix keys in Redis.
- `ping` (object): Configure ping to Redis DB.

### Buckets:

- `size` (number): is the maximum content of the bucket. This is the maximum burst you allow.
- `per_interval` (number): is the amount of tokens that the bucket receive on every interval.
- `interval` (number): defines the interval in milliseconds.
- `unlimited` (boolean = false): unlimited requests (skip take).
- `skip_n_calls` (number): take will go to redis every `n` calls instead of going in every take.

> - `elevated_limits` (object): elevated limits configuration that kicks in when the bucket is empty. Please refer to the [ERL section](##ERL-Elevated-Rate-Limits) for more details.

Reviewer: extra hash sign in link

```suggestion
- `elevated_limits` (object): elevated limits configuration that kicks in when the bucket is empty. Please refer to the [ERL section](#ERL-Elevated-Rate-Limits) for more details.
```

Author: Fixed

- `per_interval` (number): is the amount of tokens that the bucket receive on every interval.
- `interval` (number): defines the interval in milliseconds.
- `unlimited` (boolean = false): unlimited requests (skip take).
- `skip_n_calls` (number): take will go to redis every `n` calls instead of going in every take.

> - `elevated_limits` (object): elevated limits configuration that kick in when the bucket is empty. Consider it as an override that kick in when the bucket is empty.

Reviewer 2: ```suggestion
- `elevated_limits` (object): elevated limits configuration that kicks in when the bucket is empty. Consider it as an override that kicks in when the bucket is empty. It contains the following structure:
  - `size` (number): the new bucket size, for example 100
  - `per_second`: the new bucket refill
  - `erl_activation_period_mins`: for how many minutes the ERL configuration should remain active
```

Author: Actually it can have any field that the overrides can have, not only those 3. Do you think we should explicitly list all of them? Or just stating "any field you can override, can also be used here" instead?

Author: Adjusted

}
```

## ERL (Elevated Rate Limits)
ERL is a feature that allows you to define a different set of limits that kick in when the bucket is empty.

> To be able to allow its use within limitd-redis, you need to set the `allowERL` to `true`

Reviewer: The `allowERL` what? setting? environment variable? Where is that? Is this `allowERL` whatever public (I guess, it will be in the code anyway, right?).

Also, missing period at the end.

```suggestion
To be able to allow its use within limitd-redis, you need to set the `allowERL` field in the `take` request to `true`.
```

Author: It's gonna be passed to [take](https://github.com/ecita/limitd-redis/blob/93406e47e321dfb5189f2fa3a40e4ec5ab3948ba/lib/db.js#L199) function call as one of `params` - it is to allow a breaking glass mechanism that can disable the whole ERL flow quickly - i.e. via mechanisms like feature flags.

Author: I'll add some more verbose explanation on why it exists 👍

Reviewer: no need for the verbose explanation. It's just that when you read the README you have no idea where is that `allowERL` thingie that should be changed.

Author: I've expanded the readme quite a bit, please take a look and let me know if anything is still unclear

You can configure elevated limits inside your type definitions as follows:

```js
buckets = {
  ip: {
    size: 10,
    per_second: 5,
    elevated_limits: {
      size: 100, // new bucket size. already used tokens will be deducted from current bucket content upon ERL activation.
      per_second: 50, // new bucket refill ate

>       erl_activation_period_mins: 5, // for how long the ERL configuration should remain active

Reviewer 2: ```suggestion
      erl_activation_period_seconds: 300, // for how long the ERL configuration should remain active
```
I'd suggest using the same time unit (second) if possible for 2 reasons: consistency and reducing the minimum time unit allowed. 
Even if the minimum time for ERL could be 5 min or 15 min, I think it could be interesting to still work at seconds in the code.

Author: AFAIR minutes were used because this is the unit of activation periods that were going to be exposed as a part of quota offerings for ERL, so the bucket definitions can be easier to read.

Author: Changed to seconds

To be able to allow its use within limitd-redis, you need to:
1. call the `take` method with the parameter `allowERL` set to `true`.
2. pass the `erlIsActiveKey` parameter with the identifier of the ERL activation for the bucket. This works similarly to the `key` you pass to `limitd.take`, which is the identifier of the bucket; however it's used to track the ERL activation for the bucket instead.
3. make sure that the bucket definition has ERL configured.

You can configure elevated limits inside your bucket definitions as follows:

```js
buckets = {
  ip: {
    size: 10,
    per_second: 5,
    elevated_limits: {
      size: 100, // new bucket size. already used tokens will be deducted from current bucket content upon ERL activation.
      per_second: 50, // new bucket refill rate. You can use all the other refill rate configurations defined above, such as per_minute, per_hour, per_interval etc.
      erl_activation_period_seconds: 300, // for how long the ERL configuration should remain active once activated.
    }
  }
}
```

## Breaking changes from `Limitdb`

* Elements will have a default TTL of a week unless specified otherwise.

## TAKE

```js

> limitd.take(type, key, [count], erlActiveKey, allowERL (err, result) => {

Reviewer: are `erlActiveKey` and `allowERL` mandatory? I thought they were optional (reading from description of the fields below).

Does it make sense that only one of those is provided? maybe we can say that if `erlActiveKey` is passed then `allowERL` is implied?

In any case, missing a comma between `allowERL` and the function.

```suggestion
limitd.take(type, key, [count], [erlActiveKey], [allowERL], (err, result) => {
```

Reviewer: Are these the real parameters `take` take? In `lib/client.js` (which I'm not really sure if that is `limitd.take(...)` we have:

```
take(type, key, opts, cb) { ... }
```

So maybe it should be written as

```suggestion
limitd.take(type, key, { [count], [erlActiveKey], [allowERL]}, (err, result) => {
```

?

Author: That's likely. Also found an [example](https://github.com/auth0/api2/blob/bdabb220e96af8e03965f0d20a0b95ea51a25b0d/api/anomaly/blocks/ips/lib/anomaly-detection.js#L42) that passes `configOverride` together with `count`.  Updated accordingly.
```
anomalyDetectionLimitrClient.take(bucketName, key, { configOverride, count: 0 }, (err, response)
```
---

> are erlActiveKey and allowERL mandatory? I thought they were optional

`allowERL` is optional. `erlIsActiveKey` is mandatory ONLY IF you are passing a bucket that has an `elevated_limits` section in its definition.

-  `key`: the identifier of the bucket.
-  `count`: the amount of tokens you need. This is optional and the default is 1.
-  `configOverride`: caller-provided bucket configuration for this operation
-  `erlIsActiveKey`: (string) the identifier of the ERL activation for the bucket. Must be passed for buckets that have ERL configured.

> -  `allowERL`: (boolean = false) if set to true, the ERL feature will be allowed to be used. Default is false.

Reviewer: Can you update the example for `limitd.take(...)` in line 140?

-  `key`: the identifier of the bucket.
-  `count`: the amount of tokens you need. This is optional and the default is 1.

> -  `configOverride`: caller-provided bucket configuration for this operation

Reviewer: this is missing in line 148 (not related to this PR per se but... 🙄 )

> -  `erlIsActiveKey`: (string) the identifier of the ERL activation for the bucket. Must be passed for buckets that have ERL configured.

Reviewer: This is shown as `erlActiveKey` (missing `Is`) in line 148. Make sure both are in sync.
